meetings/working-groups/field-keyword/FK-2024-05-26.md

```diff
@@ -0,0 +1,255 @@
+
+```c#
+string Prop { get => field ?? ""; } // or
+string Prop { get => field ??= ComputeInitialValue(); }
+```
+
+You would get a warning in each constructor that the backing field `_field` was not initialized through all codepaths.  This would then require the user to have to suppress the warning, or write something like the following:
+
+```c#
+string Prop { get => field ??= ComputeInitialValue(); } = null!; // or
+
+// in constructor:
+this.Prop = null!;
+```
+
+So, while simple, migrating a trivial, idiomatic case like:
+
+```c#
+string? _prop;
+string Prop { get => _prop ??= ComputeInitialValue(); }
+```
+
+would involve a non-trivial transformation, with potentially annoying suppressions.
+
+### Approach A2. Control backing field warnings with attributes.
+
+With this approach:
+
+```c#
+[field: MaybeNull] string Prop { ...field... }
+
+// Would be the same as
+[MaybeNull] string _field;
+string Prop { ..._field... }
+```
+
+This has the benefit of now both suppressing any warnings about the field being `null` in any constructor, while also providing appropriate warnings if `field` was not used safely.  For example:
+
+```c#
+[field: MaybeNull] string Prop { get => field; } // will appropriately warn
+[field: MaybeNull] string Prop { get => field ??= ComputeValue(); } // will appropriately not warn.
+```
+
+Note: the above approach would also allow for the case where the setter wanted to reset the value back to `null` by using `AllowNull`.  Specifically:
+
+```c#
+// will appropriately not warn.
+[field: MaybeNull, AllowNull] string Prop { get => field ??= ComputeValue(); set => field = value is "" ? null : value.Trim(); } 
+```
+
+Like with `A1` though, this whole approach comes with the annoying nuisance of having to annotate the field in this fashion to avoid warnings in the constructor. Even if it seems like something the language/compiler could understand.
+
+
+## Approach B.  Backing field has the nullable version of the property type.
+
+### Approach B1. Use existing rules about initializing an auto-prop from a constructor.
+
+With this approach:
+
+```c#
+string Prop { ...field... }
+
+// Would be the same as the following for the purpose of analyzing the accessors
+string? _field;
+string Prop { ..._field... }
+```
+
+However, like with auto-properties today, one would still be required to initialize the property in the constructor, even if not necessary.  So:
+
+```c#
+public C()
+{
+    // would warn, the auto-prop is not considered initialized
+    Console.WriteLine(this.Prop);
+}
+```
+
+Whereas:
+
+```c#
+public C()
+{
+    this.Prop = "";
+    // would no longer warn, the auto-prop is considered initialized
+    Console.WriteLine(this.Prop);
+}
+```
+
+Note that this analysis is not accurate.  Consider a setter that resets the backing field back to null:
+
+```c#
+string Prop { get; set => field = value is "" ? null : value }
```

Comment:
That's part of the open pros/cons of this approach.  because if it warns, then you have no ability to suppress that warning, making this even more painful.  The idea here would be to say that the property is fine presuming the constructors leave it in an assigned state at their end.  because from *that* point onwards, any reads will be reading from an initialized value.

(this is inherently why some of these approaches are simple, but not great, and why we have choices here in what path to go).

Reply:
What rule though makes it not warn? This section only details that the backing field is `string?` and by such that should warn on the `get`. I agree it's inconvenient but it's hard to evaluate this until we understand what rules are happening here.

Reply:
Because it is known that the constructors are required to initialize the property to a non-null state (since they would warm otherwise), the field is assumed to be in a non-null state in the getter. 

No interprocedural analysis is done here. It's just a weak heuristic to try to get a better ratio of false positives and false negatives.

Reply:
> What rule though makes it not warn?

This is really not intended to be a full spec.  These are just the very right notes for our discussion.  Designs, initial specs, and rough costs are our next steps to flesh this out and bring to ldm to determine a direction to go.

Reply:
> Because it is known that the constructors are required to initialize the property to a non-null stat

The sample though explicitly sets `field` to `null` in the `set`. So no matter what the ctor does the value can be `null`.

Reply:
Correct.  Hence why this is a weak heuristic with potential false negatives.  That's the point i'm trying to make here.  It makes some things more pleasant, but at potential downsides.  If there weren't downsides, we'd just say "let's do this and this alone", and not have other options to consider :)

This is why we want to pursue more into all of these, coming up with all the cases we know we and customers want, and identifying how all of these options do against all of them :) 

But that's followup work.  The point here is simply to say here's an *incredibly brief* outline of 5 options we talked about,  and our next steps is a thorough examination which we can then use to bring option and recommendations to the LDM.

Reply:
And, to be clear, in no way were we saying "option B1 is good, and we'll be pushing on that".  Rather it was "hey... here's a potential path we can take that is likely low cost.  we shoudl consider it and understand what it might be good or bad at".

In the discussion there were strong negative reactions to many of these options by different individuals.  But we're not ready to eliminate any of them yet until we have a better understanding.  And, we'll then take the viable options to LDM to make a real call on what path to actually go down.

Reply:
> Hence why this is a weak heuristic with potential false negatives. T

But you haven't defined the heuristic hence I have no way to reason about this. The part you have defined does not match up with the behavior you spec'd. 

I'm not asking for a full spec here but I see no explanation as to how this works the way you describe.

Reply:
these are intended to be rough notes on the idea we had.  it is not intended to be something that even holds up to scrutiny.  already, many bugs and issues have been found.  All i was trying to do was document what we were thinking about so we can come back and continue on this, putting our heads together to flesh our a real design for each.  It may end up that some of these are non-sensical and get entirely disregarded.